app/src/main/java/com/realityexpander/tasky/agenda_feature/data/repositories/agendaRepository/agendaRepositoryImpls/AgendaRepostoryImpl.kt

```diff
@@ -39,24 +44,40 @@ class AgendaRepositoryImpl @Inject constructor(
             supervisorScope {
                 launch(Dispatchers.IO) {
                     try {
-                        send(eventRepository.getEventsForDay(dateTime)) // send stale local data
+                        // send stale local data
+                        send(eventRepository.getEventsForDay(dateTime) +
+                                taskRepository.getTasksForDay(dateTime)
+                        )
 
                         // Get fresh data
                         val result = agendaApi.getAgenda(dateTime)
 
-//                        eventRepository.clearEventsForDayLocally(dateTime) // clear local data // todo add this
+//                        eventRepository.clearEventsForDayLocally(dateTime) // clear local data // todo add this?
+                        // Insert fresh data into db
                         result.events.forEach { event ->
-                            // Insert fresh data into db
                             val result2 =
                                 eventRepository.upsertEventLocally(event.toDomain())
                             if(result2 is ResultUiText.Error) {
                                 throw IllegalStateException(result2.message.asStrOrNull())
                             }
                         }
 
-                        // Emit the current state from DB
-                        eventRepository.getEventsForDayFlow(dateTime).collect { events ->
-                            send(events)
+//                        taskRepository.clearTasksForDayLocally(dateTime) // clear local data // todo add this?
+                        // Insert fresh data into db
+                        result.tasks.forEach { task ->
+                            val result2 =
+                                taskRepository.upsertTaskLocally(task.toDomain())
+                            if(result2 is ResultUiText.Error) {
+                                throw IllegalStateException(result2.message.asStrOrNull())
+                            }
+                        }
+
+                        taskRepository.getTasksForDayFlow(dateTime).combine(
+                            eventRepository.getEventsForDayFlow(dateTime)
+                        ) { tasks, events ->
+                            events + tasks // + reminders
+                        }.collect { agendaItems ->
+                            send(agendaItems)
```

Comment:
## FOLLOW UP QUESTION 1

When I attempt to use a `flow` here, i get this error message and im not sure exactly why. I have googled it but nothing shows up about it.

```
java.lang.IllegalStateException: Flow invariant is violated:
Emission from another coroutine is detected.
Child of StandaloneCoroutine{Active}@f38a6ac, expected child of StandaloneCoroutine{Active}@2ef8f75.
FlowCollector is not thread-safe and concurrent emissions are prohibited.
To mitigate this restriction please use 'channelFlow' builder instead of 'flow'
kotlinx.coroutines.flow.internal.SafeCollector_commonKt$checkContext$result$1.invoke(SafeCollector.common.kt:68)
kotlinx.coroutines.flow.internal.SafeCollector_commonKt$checkContext$result$1.invoke(SafeCollector.common.kt:25)
```
This is why I switched to `channelFlow` and cant use `emit` or `emitAll`.

I am stumped to why this is happening. What about my implmentation is wrong?

Reply:
> but couldn't it be better to have one function which just returns the combined flow from the DAO and one function which updates the cache? As soon as the cache gets updated, the combine flow will emit.

OK, I will refactor to use separate `Update` and `Get` functions.

Reply:
Yes I think you're not allowed to emit from a different coroutine (what happened in your `launch` block). However what you can do is to use `async` instead to still execute stuff in parallel, but you get the result in the flow-coroutine and emit it there. That shouldn't cause any issues.
I think when emitting inside a coroutine, the flow has issues detecting when it's finished.

Reply:
Yes that was the issue, the separate co-routine. I removed it and `emit` and `emitAll` work fine now.